### ISSUE TITLE
LoRA is not wired into the training path — nothing trains through LoRALayer/LoRAModel

### ISSUE THREAD

**Issue by MBrekhof** [Integral2u/SharpMind#23]
Following on from #13, where we settled that accelerators arrive as plugins: I hit a wall trying to bring the GPU training engine across, and the wall is LoRA rather than anything GPU-specific. Raising it here rather than sending a patch, because the fix deletes classes you wrote.

## The problem

`SharpMind.Training/LoRA/` ships `LoRAConfig`, `LoRALayer`, `LoRAModel`, `LoRAAttention` and `LoRAFFN` — but nothing trains through them. On `b1766a9`:

| | mentions of "lora" |
|---|---|
| `SharpMind.Model/Layers/TrainingLinearLayer.cs` | **0** |
| `SharpMind.Training/Autograd/BackpropEngine.cs` | **0** |

And `LoRAModel` / `LoRAParameters()` are referenced by nothing outside the `LoRA/` folder itself — not `TrainLoop`, not the CUI, not the model layer.

`LoRALayer.Forward(input, frozenWeights)` computes `frozen(x) + scale * B(A(x))` correctly, but it is a **parallel object** sitting beside the model rather than inside it. The autograd engine walks `Transformer`'s layers; it never sees a `LoRALayer`, so no gradient ever reaches `A` or `B`. `LoRAModel.TrainableRatio()` reports a number that describes a model that cannot be trained.

So today `TrainLoop` can only do full fine-tuning: every parameter trainable, optimiser state for all of them.

## Why I care right now

The GPU plugin's whole value is **LoRA fine-tuning on device** — freeze the base weights, keep only the small adapters live, and a 0.5B model fits comfortably on one card. Measured on a rented RTX 3090: 553 ms/step for 0.5B LoRA r=8 at batch 4 x seq 64, against 8,420 ms for the CPU engine on the same machine, with loss curves matching to four decimal places.

Without LoRA wired into the training path there is nothing for that engine to accelerate. It is also what makes fine-tuning viable on consumer hardware at all — full fine-tuning of a 1.5B model needs optimiser state that will not fit on most cards.

## What I think the fix is

Put the adapter **inside** the layer the autograd engine already traverses, rather than beside it:

```csharp
// TrainingLinearLayer
public bool  HasLoRA   => _loraA is not null;
public int   LoRARank  => _loraA?.Shape.Cols ?? 0;
public float LoRAScale => _loraScale;
public void  EnableLoRA(int rank, float scale, Random rng);
```

`BackpropEngine` then routes gradient to `A`/`B` and skips the frozen base weight. That also unlocks a real optimisation on the backward pass: a frozen head needs only `dFinalNormOut`, one blocked matmul against `headᵀ`, with no weight gradient at all. At the 0.5B training shape that is 3.1 s for the trainable kernel's `dInput + dW` versus 0.12 s to transpose plus 1.45 s for the matmul.

The cost is that `LoRALayer`, `LoRAAttention` and `LoRAFFN` stop having a role, and `LoRAModel` becomes a thin descriptor over the layers that have adapters enabled.

## What I have

A working implementation, currently on my branch: `TrainingLinearLayer` +97, `BackpropEngine` +196, `LoRAModel` rewritten, those three classes removed, and `LoRATests` reworked against the new shape (410 lines changed) — it trains, and the GPU engine sits on top of it.

I did not want to open that as a pull request unannounced. It deletes three of your classes, and #13 taught me that is a conversation worth having first.

So: does this direction suit you? I am happy to send it as one PR, or to split it — adapters in `TrainingLinearLayer` first, then the `BackpropEngine` routing, then the `LoRAModel` cleanup — whichever is easier to review. And if you would rather keep `LoRALayer` as a public forward-only API and have me wire the adapters underneath it instead, say so and I will shape it that way.

**Comment by Integral2u** [comment]
Yeah I assumed there will be some bigger changes that need to be made, hopefully once everything is patched back up on the release of 1.0.5.0 users will need to make minimal code correction.  But right now it's better to rip off the band aid.

**Comment by Integral2u** [comment]
resolved with PR #26
